avroregistry/registry.go

```diff
@@ -190,25 +190,38 @@ func (r *Registry) doRequest(req *http.Request, result interface{}) error {
 		resp, err := http.DefaultClient.Do(req)
 		if err != nil {
 			if !attempt.More() || !isTemporaryError(err) {
-				return err
+				return &UnavailableError{err}
 			}
 			continue
 		}
 		err = unmarshalResponse(req, resp, result)
 		if err == nil {
 			return nil
 		}
-		if !attempt.More() {
-			return err
-		}
-		if err, ok := err.(*apiError); ok && err.StatusCode/100 != 5 {
-			// It's not a 5xx error. We want to retry on 5xx
+		if err, ok := err.(*apiError); ok {
```

Comment:
Trying to refactor this:

```
	for attempt.Next() {
		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			if !attempt.More() || !isTemporaryError(err) {
				return &UnavailableError{err}
			}
			continue
		}
		err = unmarshalResponse(req, resp, result)
		if err == nil {
			return nil
		}
		if apiErr, ok := err.(*apiError); ok {
			// We want to retry on 5xx
			// errors, because the Confluent Avro registry
			// can occasionally return them as a matter of
			// course (and there could also be an
			// unavailable service that we're reaching
			// through a proxy).
			if err.StatusCode/100 == 5 {
				return apiErr
			} else {
				err = &UnavailableError{apiErr}
			}
		}

		if !attempt.More() {
			return err
		}
	}
```

Reply:
Added it with a case where a proxy returns 5XX status and non-json body and it's not an [apiError anymore](https://github.com/heetch/avro/pull/133/commits/4f1eb1e71177631c19f785dcf62bcb5c0fdfe40c).

Reply:
:up: